base/compiler/typeinfer.jl

```diff
@@ -1,8 +1,8 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 # Tracking of newly-inferred MethodInstances during precompilation
-const track_newly_inferred = RefValue{Bool}(false)
 const newly_inferred = MethodInstance[]
+const consume = :monotonic
```

Comment:
@tkf is this even accurate though? `consume` ordering implies it does data dependency tracking, which this does not do. What we want is only `monotonic`+depends as used in the linux kernel (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0124r7.html)

Reply:
Should I then define this as `depends` instead of `consume`?

Reply:
The linux kernel model defines `depends` as a no-op, and defines that `monotonic` is sufficient for all of our uses

Reply:
That's an implementation and not a definition. There is a formal definition of `monotonic`. It doesn't talk about dependencies.

Also, the implementation has to use `monotonic` and volatile, IIUC. This is probably not required for how it's used in this patch, though.

Reply:
@pchintalapudi  Actually, this PR is probably not the right place to discuss this. Please feel free to revert my suggestion and use literal `:monotonic` everywhere. I probably should open this myself sometime later.

Reply:
The use of `volatile` is orthogonal and forces forward progress for READ_ONCE, which otherwise is not necessarily specified in the C11 memory model, and is not relevant here. All we exactly care is that the linux memory model defines that monotonic is correct here and sufficient (because it defines that `depends` is a no-op).

The linux kernel has also, thus far, rejected the C11 memory model as the model definition of `monotonic` does not match the usual hardware behavior. That is why I suggested we will follow the linux memory model, not C11 here.

Reply:
> we will follow the linux memory model, not C11 here

By "here", do you mean in the implementation of the Julia compiler or in the Julia language?  I gave up on complaining when I see LKMM in the compiler and the runtime (in fact, the only point I'm suggesting in this PR is to be explicit for cases when/how it uses LKMM) but I'll still strongly oppose the idea of using LKMM in the Julia language :)

Reply:
The compiler, runtime, and language are all implemented in the Julia language though 😂

Reply:
If you pick up a "non-default" memory model like LKMM for a given piece of code, you are not using the language that is typically associated with the syntax of the program you are writing.  In this sense, Linux is not written in C.  Likewise, I didn't mind if the Julia compiler is written in `*.jl` files but does not use Julia language.  All I'm suggesting here is to clearly mark the places that do not use the Julia language.

(In this comment, I'm assuming that memory ordering semantics is equivalent to C++, of course. If you specify the semantics of `:monotonic` as something different, it may be possible to use LKMM in the Julia language.)